src/robot/parsing/model/statements.py

```diff
@@ -971,26 +971,35 @@ class WhileHeader(Statement):
     type = Token.WHILE
 
     @classmethod
-    def from_params(cls, condition, indent=FOUR_SPACES, separator=FOUR_SPACES, eol=EOL):
+    def from_params(cls, condition, limit=None, indent=FOUR_SPACES,
+                    separator=FOUR_SPACES, eol=EOL):
         tokens = [Token(Token.SEPARATOR, indent),
                   Token(cls.type),
                   Token(Token.SEPARATOR, separator),
-                  Token(Token.ARGUMENT, condition),
-                  Token(Token.EOL, eol)]
+                  Token(Token.ARGUMENT, condition)]
+        if limit:
+            tokens.extend([Token(Token.SEPARATOR, indent),
+                           Token(Token.OPTION, limit)])
+        tokens.append(Token(Token.EOL, eol))
         return cls(tokens)
 
     @property
     def condition(self):
-        values = self.get_values(Token.ARGUMENT)
-        if len(values) != 1:
-            return ', '.join(values) if values else None
-        return values[0]
+        return ', '.join(self.get_values(Token.ARGUMENT))
+
+    @property
+    def limit(self):
+        value = self.get_value(Token.OPTION)
+        return value.replace('limit=', '') if value else None
 
     def validate(self, context):
-        conditions = len(self.get_tokens(Token.ARGUMENT))
-        if conditions == 0:
+        values = self.tokens[1:]
+        if len(values) == 0:
             self.errors += ('WHILE must have a condition.',)
-        if conditions > 1:
+        if len(values) == 2 and values[1].type != Token.OPTION:
+            self.errors += (
+                f"Second WHILE loop argument must be 'limit', got {values[1]}.",)
```

Comment:
I don't think we anymore need to validate limit at all. Lexing works so that it can only be at the end and must always start with `limit=` anyway.